kmath-core/src/commonMain/kotlin/space/kscience/kmath/misc/sorting.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2018-2021 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+ */
+
+package space.kscience.kmath.misc
+
+import kotlin.comparisons.*
+import space.kscience.kmath.data.ColumnarData
+import space.kscience.kmath.expressions.Symbol
+import space.kscience.kmath.misc.UnstableKMathAPI
+import space.kscience.kmath.structures.Buffer
+import space.kscience.kmath.structures.indices
+
+/* TODO: Unify access API of indexed-based data structures. [Buffer], [ColumnarData] or other concepts like tables/dataframes
+ * use a key/value pairing to access/reference data values. In case of buffer, it is an implicit key in the form of a 
+ * monotonic integer value. For SQL tables, most tables contains a "primary key" column, whose values (of arbitrary type)
+ * are ensured unique and can be used as an index/identifier to retrieve an associated record/value.
+ * Note that the simple concept of key-value pairing can allow to define a wide range of operations like subset, concatenation,
+ * permutations (shuffle, sort), etc.
+ *
+ * Defining an API for this could help factorizing code, and more importantly, define a unified set of properties/operators over 
+ * a wide range of data structures. 
+ */
+private interface Registry<K, V> {
```

Comment:
Why do we need a new type? It seems to behave as a regular Map.

Reply:
I have, indeed, done over-engineering. I will rework the PR to make it really simple / little.

While coding a simple solution for buffers, I've thought about how it could be made more generic, but I think I've gone too far. My thought is that any data-structure providing get() operator is (implicitely for buffers or explicitely for maps) an indexed data structure. To allow using the same permSort() function over different data structures, I've tried to model what is strictly necessary in term of key/index permutation. But it's far beyond this PR goal (even maybe beyond the goal of the library), so I'll remove it.